component/authenticator/src/webapp/voiceMobile.jsp

```diff
@@ -0,0 +1,203 @@
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied.  See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  --%>
+
+<%@ page import="org.owasp.encoder.Encode" %>
+<%@ page import="org.wso2.carbon.identity.mgt.endpoint.util.IdentityManagementEndpointUtil" %>
+<%@ page import="org.wso2.carbon.identity.application.authentication.endpoint.util.Constants" %>
+<%@ page import="java.io.File" %>
+<%@ page import="java.util.Map" %>
+<%@ page import="org.apache.commons.lang.StringUtils"%>
+<%@ page language="java" contentType="text/html; charset=UTF-8" pageEncoding="UTF-8" %>
+<%@ include file="includes/localize.jsp" %>
+<%@ page import="static java.util.Base64.getDecoder" %>
+<%@ page import="org.wso2.carbon.identity.authenticator.voiceotp.VoiceOTPConstants" %>
+<%@ page import="org.apache.commons.lang.StringUtils" %>
+<%@ taglib prefix="layout" uri="org.wso2.identity.apps.taglibs.layout.controller" %>
+
+<jsp:directive.include file="includes/layout-resolver.jsp"/>
+
+<%
+    request.getSession().invalidate();
+    String queryString = request.getQueryString();
+    Map<String, String> idpAuthenticatorMapping = null;
+    String mobileRegex = null;
+    boolean validateMobileNumberFormat = false;
+    String mobileRegexPolicyValidationErrorMessage = null;
+    if (request.getAttribute(Constants.IDP_AUTHENTICATOR_MAP) != null) {
+        idpAuthenticatorMapping = (Map<String, String>) request.getAttribute(Constants.IDP_AUTHENTICATOR_MAP);
+    }
+    if (StringUtils.isNotBlank(request.getParameter(VoiceOTPConstants.MOBILE_NUMBER_REGEX_PATTERN))) {
+        mobileRegex = new String(getDecoder().decode(request.getParameter(VoiceOTPConstants.MOBILE_NUMBER_REGEX_PATTERN)));
+        validateMobileNumberFormat = true;
+    }
+    if (StringUtils.isNotBlank(request.getParameter(VoiceOTPConstants.MOBILE_NUMBER_PATTERN_POLICY_FAILURE_ERROR_MESSAGE))) {
+        mobileRegexPolicyValidationErrorMessage = new String(getDecoder().decode(request.getParameter(VoiceOTPConstants.MOBILE_NUMBER_PATTERN_POLICY_FAILURE_ERROR_MESSAGE)));
+    }
+
+    String errorMessage = IdentityManagementEndpointUtil.i18n(resourceBundle,"error.retry");
+    boolean authenticationFailed = false;
+
+    if (Boolean.parseBoolean(request.getParameter(Constants.AUTH_FAILURE))) {
+        authenticationFailed = true;
+
+        if (request.getParameter(Constants.AUTH_FAILURE_MSG) != null) {
+            errorMessage = request.getParameter(Constants.AUTH_FAILURE_MSG);
+
+                if ("authentication.fail.message".equalsIgnoreCase(errorMessage)) {
+                    errorMessage = "Authentication Failed! Please Retry";
+                }
+
+                if (StringUtils.isNotBlank(request.getParameter("authFailureInfo"))) {
+                    errorMessage = request.getParameter("authFailureInfo");
```

Comment:
If we set the query parameters directly in the error message (lines 59 and 66), we won't be able to localize the content on the frontend. It's better to use the i18n key as a parameter, check if it exists in the resource bundle, and render it. If the key is not found, you can display the default message.